torchtext/transforms.py

```diff
@@ -189,3 +193,350 @@ def forward(self, input: Any) -> Any:
+
+        Equivalent to:
+        min(pairs, key = lambda pair: self.bpe_merge_ranks.get(pair, float('inf')))
+        """
+        best_pair: str = pairs[0]
+        best_rank: int = self.bpe_merge_ranks.get(best_pair, self.inf)
+
+        for pair in pairs[1:]:
+            rank: int = self.bpe_merge_ranks.get(pair, self.inf)
+            if rank < best_rank:
+                best_pair = pair
+                best_rank = rank
+        return best_pair
+
+    @torch.jit.export
+    def bpe(self, token_list: List[str]) -> List[str]:
+        """Return a list of bpe tokens.
+
+        Given a list of input tokens, keep finding the best bpe merge and
+        generate a new list of tokens until
+        1) token list size reduced to 1 or
+        2) can't find bpe merge
+
+        :param token_list: a list of encoded bytes, each represented by an unicode.
+                For example: ['a', 'w', 'e', 's', 'o', 'm', 'e']
+        :type input: List[str]
+        :return: A list of bpe tokens generated by greedy search.
+                For example: ["aw", "esome"]
+        :rtype: List[str]
+        """
+        pairs: List[str] = self.get_pairs(token_list)
+
+        if len(pairs) == 0:
+            return token_list
+
+        while True:
+            bigram: str = self.find_best_pair(pairs)
+            if bigram not in self.bpe_merge_ranks:
+                break
+
+            # Finding all indexes that token_list[i] == first and
+            # token_list[i+1] == second.
+            # After the loop, new token list will be
+            # 1) first + second pair
+            # 2) all the other tokens in the original token list
+            #
+            # For example: first="a" second="w" and token_list =
+            # ["a", "w", "some", "a", "w", "e"]
+            # Result: new_token_list = ["aw", "some", "aw", "e"]
+            first, second = bigram.split(self.merge_seperator)
+            new_token_list: List[str] = []
+            i: int = 0
+            while i < len(token_list):
+                j = self.list_str_index(token_list, first, i)
+                if j != -1:
+                    new_token_list.extend(token_list[i:j])
+                    i = j
+                else:
+                    new_token_list.extend(token_list[i:])
+                    break
+
+                if (
+                    token_list[i] == first
+                    and i < len(token_list) - 1
+                    and token_list[i + 1] == second
+                ):
+                    new_token_list.append(first + second)
+                    i += 2
+                else:
+                    new_token_list.append(token_list[i])
+                    i += 1
+
+            token_list = list(new_token_list)
+            if len(token_list) == 1:
+                break
+            else:
+                pairs = self.get_pairs(token_list)
+
+        return token_list
+
+    @torch.jit.export
+    def byte_encode(self, token: str) -> List[str]:
```

Comment:
Is there a test to verify that the unicode input/output patterns are sam for eager mode execution and scripted execution?

I guess that the logic branch at `torch.jit.is_scripting` is to handle the difference of string handling between Python and C++, but it would be more assuring to have that in test.

In the past, audio had an issue where string I/O causes different results in eager and scripted executions. The root cause is the comparison and it does not apply to this method, but I think it
s safer to have such test.

https://github.com/pytorch/audio/pull/1601#issuecomment-866972482

Reply:
The currently added unit tests actually test this. If you see the test, we run one version of tests with the nn.Module while the other is run after scripting the module. The test has input cases with special unicode characters as well for completeness. We expect same output in both cases.  Furthermore, the scripted implementation is also something that is well tested internally.